### PR TITLE
Simplify correlation IDs

### PR DESCRIPTION
**Pull request by adriansmares**
#### Summary

This PR proposes a rework of the correlation ID system, by introducing the concept of correlation ID classes and the removal of long lived correlation IDs.

The idea behind the correlation ID classes is that having multiple correlation IDs relating to the same message is superfluous. Any extra correlation ID outside of the source component correlation ID (`gs`/`ns`/`pba` for uplinks, and `as`/`ns`/`pba` for downlinks) is not adding any information outside of 'I've passed this component'. But that is not really information - the traffic path is fixed outside of the entry point: you always go (GS/PBA) -> NS (-> AS) - there isn't any magic path that you are learning from the correlation IDs.

Correlation ID classes make prefixes exclusive one with another - if an uplink originates from the GS, only the `gs:uplink:*` correlation ID will be present, and the NS/AS won't append theirs too. This logic is meant to lower the bandwidth and storage usage both in-between components and with the outside world.

The removal of the long lived correlation IDs refers to the removal of the correlation IDs tied to a connection - the idea here is that for an uplink message it is already possible to trace back the source of the message by looking at the RX metadata, and that is already more useful than a unique identifier for the connection.

The use cases supported by the correlation ID system are still doable - you can still trace the entry point of a message, and all of the correlated events, but the contents are now lighter.

#### Changes

- Remove connection correlation IDs.
- Add correlation ID classes.
- Support ignoring certain RPCs from the correlation ID injection process.

#### Testing

Manual testing on `staging1` can cover the following subjects:

- Add an OTAA end device and do a full - join -> first uplink - > application downlink flow.
- Uplink related events should not have any `rpc:` correlation IDs, and only the `gs:uplink:` correlation ID should be present in general - correlation IDs like `ns:uplink:` or `as:up:` should not be visible.
- Downlink related events from the application level should have the `as:downlink:` correlation ID, while network level downlinks (like the initial `LinkADRReq` downlinks) should have the `ns:downlink:` correlation ID. No other component downlink correlation IDs should be visible. The transmission correlation ID (`ns:transmission:`) can be present.

Examples of how small the correlation IDs are now:

- Uplink message:
image

- Application layer downlink:
image

- Application layer transmission acknowledgement:
image

- Network layer downlink:
image

- Network layer transmission acknowledgement:
image

##### Regressions

This change affects the correlation IDs reported in events and upstream messages, by removing certain correlation IDs in effect. Applications generally do not depend on the stack assigned correlation IDs, and application provided correlation IDs are unaffected as long as they did not use stack prefixes.

#### Notes for Reviewers

This is mainly a proposal PR, so please let me know your thoughts.

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [x] The steps/process to test this feature are clearly explained including testing for regressions.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: pkg/applicationserver/io/grpc/grpc.go [hunk 1]
```diff
@@ -213,9 +214,18 @@ func (s *impl) GetMQTTConnectionInfo(ctx context.Context, ids *ttnpb.Application
 var errPayloadCryptoSkipped = errors.DefineFailedPrecondition("payload_crypto_skipped", "payload crypto skipped")
 
 func (s *impl) SimulateUplink(ctx context.Context, up *ttnpb.ApplicationUp) (*emptypb.Empty, error) {
-	if err := rights.RequireApplication(ctx, up.EndDeviceIds.ApplicationIds, ttnpb.Right_RIGHT_APPLICATION_TRAFFIC_UP_WRITE); err != nil {
+	if up.Simulated {
+		// Traffic already marked as simulated is discarded, in order to avoid traffic loops created by injecting
+		// simulated traffic which is then piped into the simulate uplink endpoint itself.
+		return ttnpb.Empty, nil
+	}
```
Comment (johanstokking) [inline pkg/applicationserver/io/grpc/grpc.go]: This is an unrelated behavioral change. I don't get the comment. What are we trying to avoid, what would this traffic loop look like? Where is AS calling `SimulateUplink()` itself?

**Review by johanstokking — COMMENTED** [review]
The verbosity in correlation IDs stems from the early pre-release V3 days. I wanted to be able to find related context from an event, i.e. the gateway's connection (with IP and protocol), and from there find other events. It would be used for generic grouping and filtering. RX metadata is not an alternative for that because it's about a specific use case and doesn't include the connection lifetime. Anyway, I don't think we have an actual use case for working with `gs:conn`, so I'm okay removing it for now and we may reintroduce it later.

File: pkg/applicationserver/io/grpc/grpc.go (hunk 1 above)
Comment (adriansmares) [reply pkg/applicationserver/io/grpc/grpc.go]: The request type for `SimulateUplink` is `ApplicationUp`, and it has the HTTP binding `/as/applications/{end_device_ids.application_ids.application_id}/devices/{end_device_ids.device_id}/up/simulate`. Webhook messages from TTS also have the request body type `ApplicationUp`.

It is possible to create a webhook which uses `/as/applications/{end_device_ids.application_ids.application_id}/devices/{end_device_ids.device_id}/up/simulate` as an endpoint, and create a loop with a single simulated message - the stack will simulate the uplink, send it to the webhooks, the webhooks call `SimulateUplink` and the process continues forever. It is even possible to make the AS generate more messages by registering two webhooks to the `SimulateUplink` endpoint - you then start the loop by simulating one message, the fanout creates two calls to `SimulateUplink`, and the cycle continues - you get thousands of inflight messages.

Historically the reason why this process actually ends is because on every RPC we would append a new correlation ID, and the correlation IDs have a maximum `repeated` size in the validation, so at some point the loop ends because the AS has appended the correlation ID too many times and the call fails in the validation phase. While this works, I would prefer to block loops explicitly instead of relying on this emergent behavior.

Comment (johanstokking) [reply pkg/applicationserver/io/grpc/grpc.go]: I see. But then we would have to comment that `simulated` is set by AS, and it is not expected to be set on incoming simulated uplink messages.

Can't we rely on a correlation ID set by AS?

Otherwise we have to update the `simulated` field in AS I think.

For clarity, we should probably return `InvalidArgument` instead of silently ignoring the message.

Comment (adriansmares) [reply pkg/applicationserver/io/grpc/grpc.go]: > But then we would have to comment that simulated is set by AS, and it is not expected to be set on incoming simulated uplink messages.

I have added an explicit comment to the `simulated` field. For the record, none of our callers (CLI/Console) have manually set it historically.

> Can't we rely on a correlation ID set by AS?

As I mentioned originally, yes we can but it is ad-hoc. I prefer to explicitly disable re-simulating traffic.

> Otherwise we have to update the simulated field in AS I think.

We do - it was always the case that the RPC marks the traffic:
https://github.com/TheThingsNetwork/lorawan-stack/blob/8c54b90132e39406d95b737808aa2d5d21979fda/pkg/applicationserver/io/grpc/grpc.go#L238-L241

> For clarity, we should probably return InvalidArgument instead of silently ignoring the message.

I agree this is better - I have added an error.

---

See the last two commits - the rest are unchanged, but I had to rebase due to conflicts.

Comment (johanstokking) [reply pkg/applicationserver/io/grpc/grpc.go]: > > Otherwise we have to update the simulated field in AS I think.
> 
> We do - it was always the case that the RPC marks the traffic:

I meant in the API doc, like you did.

**Review by johanstokking — APPROVED** [review]
